// yapi-to-typescript/lib/cjs/SwaggerToYApiServer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.SwaggerToYApiServer = void 0;

var _getPort = _interopRequireDefault(require("get-port"));

var _got = _interopRequireDefault(require("got"));

var _http = _interopRequireDefault(require("http"));

var _signalExit = _interopRequireDefault(require("signal-exit"));

var _url = _interopRequireDefault(require("url"));

var _vtils = require("vtils");

var _swaggerJsonToYApiData = require("./swaggerJsonToYApiData");

class SwaggerToYApiServer {
  constructor(options) {
    this.options = options;
    this.port = 0;
    this.swaggerJson = {};
    this.httpServer = null;
    this.yapiData = {};
  }

  async getPort() {
    if (this.port === 0) {
      this.port = await (0, _getPort.default)({
        port: 50505
      });
    }

    return this.port;
  }

  async getUrl() {
    return `http://127.0.0.1:${await this.getPort()}`;
  }

  async getSwaggerJson() {
    if ((0, _vtils.isEmpty)(this.swaggerJson)) {
      const res = await _got.default.get(this.options.swaggerJsonUrl, {
        responseType: 'json'
      });
      this.swaggerJson = res.body;
    }

    return this.swaggerJson;
  }

  async getYApiData() {
    if ((0, _vtils.isEmpty)(this.yapiData)) {
      this.yapiData = await (0, _swaggerJsonToYApiData.swaggerJsonToYApiData)(await this.getSwaggerJson());
    }

    return this.yapiData;
  }

  async start() {
    const yapiData = await this.getYApiData(); // eslint-disable-next-line no-async-promise-executor

    await new Promise(async resolve => {
      this.httpServer = _http.default.createServer(async (req, res) => {
        const {
          pathname
        } = _url.default.parse(req.url || '');

        res.setHeader('Content-Type', 'application/json');

        if (pathname.includes('/api/plugin/export')) {
          res.end(JSON.stringify(yapiData.cats.map(cat => ({ ...cat,
            list: yapiData.interfaces.filter(item => item.catid === cat._id)
          }))));
        } else if (pathname.includes('/api/interface/getCatMenu')) {
          res.end(JSON.stringify({
            errcode: 0,
            errmsg: '成功！',
            data: yapiData.cats
          }));
        } else if (pathname.includes('/api/project/get')) {
          res.end(JSON.stringify({
            errcode: 0,
            errmsg: '成功！',
            data: yapiData.project
          }));
        }

        res.end('404');
      }).listen(await this.getPort(), '127.0.0.1', () => {
        (0, _signalExit.default)(() => this.stop());
        resolve();
      });
    });
    return this.getUrl();
  }

  async stop() {
    return new Promise((resolve, reject) => {
      if (!this.httpServer) {
        resolve();
      } else {
        this.httpServer.close(err => {
          if (err) {
            reject(err);
          } else {
            resolve();
          }
        });
      }
    });
  }

}

exports.SwaggerToYApiServer = SwaggerToYApiServer;
