// yapi-to-typescript/lib/cjs/types.js
"use strict";

exports.__esModule = true;
exports.ResponseBodyType = exports.RequestFormItemType = exports.RequestQueryType = exports.RequestParamType = exports.RequestBodyType = exports.Required = exports.Method = void 0;

/** 请求方式 */
let Method;
/** 是否必需 */

exports.Method = Method;

(function (Method) {
  Method["GET"] = "GET";
  Method["POST"] = "POST";
  Method["PUT"] = "PUT";
  Method["DELETE"] = "DELETE";
  Method["HEAD"] = "HEAD";
  Method["OPTIONS"] = "OPTIONS";
  Method["PATCH"] = "PATCH";
})(Method || (exports.Method = Method = {}));

let Required;
/** 请求数据类型 */

exports.Required = Required;

(function (Required) {
  Required["false"] = "0";
  Required["true"] = "1";
})(Required || (exports.Required = Required = {}));

let RequestBodyType;
/** 请求路径参数类型 */

exports.RequestBodyType = RequestBodyType;

(function (RequestBodyType) {
  RequestBodyType["query"] = "query";
  RequestBodyType["form"] = "form";
  RequestBodyType["json"] = "json";
  RequestBodyType["text"] = "text";
  RequestBodyType["file"] = "file";
  RequestBodyType["raw"] = "raw";
  RequestBodyType["none"] = "none";
})(RequestBodyType || (exports.RequestBodyType = RequestBodyType = {}));

let RequestParamType;
/** 请求查询参数类型 */

exports.RequestParamType = RequestParamType;

(function (RequestParamType) {
  RequestParamType["string"] = "string";
  RequestParamType["number"] = "number";
})(RequestParamType || (exports.RequestParamType = RequestParamType = {}));

let RequestQueryType;
/** 请求表单条目类型 */

exports.RequestQueryType = RequestQueryType;

(function (RequestQueryType) {
  RequestQueryType["string"] = "string";
  RequestQueryType["number"] = "number";
})(RequestQueryType || (exports.RequestQueryType = RequestQueryType = {}));

let RequestFormItemType;
/** 返回数据类型 */

exports.RequestFormItemType = RequestFormItemType;

(function (RequestFormItemType) {
  RequestFormItemType["text"] = "text";
  RequestFormItemType["file"] = "file";
})(RequestFormItemType || (exports.RequestFormItemType = RequestFormItemType = {}));

let ResponseBodyType;
exports.ResponseBodyType = ResponseBodyType;

(function (ResponseBodyType) {
  ResponseBodyType["json"] = "json";
  ResponseBodyType["text"] = "text";
  ResponseBodyType["xml"] = "xml";
  ResponseBodyType["raw"] = "raw";
})(ResponseBodyType || (exports.ResponseBodyType = ResponseBodyType = {}));
