// vtils/_cjs/date/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  add: true,
  addBusinessDays: true,
  addDays: true,
  addHours: true,
  addISOWeekYears: true,
  addMilliseconds: true,
  addMinutes: true,
  addMonths: true,
  addQuarters: true,
  addSeconds: true,
  addWeeks: true,
  addYears: true,
  areIntervalsOverlapping: true,
  clamp: true,
  closestIndexTo: true,
  closestTo: true,
  compareAsc: true,
  compareDesc: true,
  daysToWeeks: true,
  differenceInBusinessDays: true,
  differenceInCalendarDays: true,
  differenceInCalendarISOWeeks: true,
  differenceInCalendarISOWeekYears: true,
  differenceInCalendarMonths: true,
  differenceInCalendarQuarters: true,
  differenceInCalendarWeeks: true,
  differenceInCalendarYears: true,
  differenceInDays: true,
  differenceInHours: true,
  differenceInISOWeekYears: true,
  differenceInMilliseconds: true,
  differenceInMinutes: true,
  differenceInMonths: true,
  differenceInQuarters: true,
  differenceInSeconds: true,
  differenceInWeeks: true,
  differenceInYears: true,
  eachDayOfInterval: true,
  eachHourOfInterval: true,
  eachMinuteOfInterval: true,
  eachMonthOfInterval: true,
  eachQuarterOfInterval: true,
  eachWeekendOfInterval: true,
  eachWeekendOfMonth: true,
  eachWeekendOfYear: true,
  eachWeekOfInterval: true,
  eachYearOfInterval: true,
  endOfDay: true,
  endOfDecade: true,
  endOfHour: true,
  endOfISOWeek: true,
  endOfISOWeekYear: true,
  endOfMinute: true,
  endOfMonth: true,
  endOfQuarter: true,
  endOfSecond: true,
  endOfToday: true,
  endOfTomorrow: true,
  endOfWeek: true,
  endOfYear: true,
  endOfYesterday: true,
  format: true,
  formatDistance: true,
  formatDistanceStrict: true,
  formatDistanceToNow: true,
  formatDistanceToNowStrict: true,
  formatDuration: true,
  formatISO: true,
  formatISO9075: true,
  formatISODuration: true,
  formatRelative: true,
  formatRFC3339: true,
  formatRFC7231: true,
  fromUnixTime: true,
  getDate: true,
  getDay: true,
  getDayOfYear: true,
  getDaysInMonth: true,
  getDaysInYear: true,
  getDecade: true,
  getHours: true,
  getISODay: true,
  getISOWeek: true,
  getISOWeeksInYear: true,
  getISOWeekYear: true,
  getMilliseconds: true,
  getMinutes: true,
  getMonth: true,
  getOverlappingDaysInIntervals: true,
  getQuarter: true,
  getSeconds: true,
  getTime: true,
  getUnixTime: true,
  getWeek: true,
  getWeekOfMonth: true,
  getWeeksInMonth: true,
  getWeekYear: true,
  getYear: true,
  hoursToMilliseconds: true,
  hoursToMinutes: true,
  hoursToSeconds: true,
  intervalToDuration: true,
  intlFormat: true,
  isAfter: true,
  isBefore: true,
  isDate: true,
  isEqual: true,
  isExists: true,
  isFirstDayOfMonth: true,
  isFriday: true,
  isFuture: true,
  isLastDayOfMonth: true,
  isLeapYear: true,
  isMatch: true,
  isMonday: true,
  isPast: true,
  isSameDay: true,
  isSameHour: true,
  isSameISOWeek: true,
  isSameISOWeekYear: true,
  isSameMinute: true,
  isSameMonth: true,
  isSameQuarter: true,
  isSameSecond: true,
  isSameWeek: true,
  isSameYear: true,
  isSaturday: true,
  isSunday: true,
  isThisHour: true,
  isThisISOWeek: true,
  isThisMinute: true,
  isThisMonth: true,
  isThisQuarter: true,
  isThisSecond: true,
  isThisWeek: true,
  isThisYear: true,
  isThursday: true,
  isToday: true,
  isTomorrow: true,
  isTuesday: true,
  isValid: true,
  isWednesday: true,
  isWeekend: true,
  isWithinInterval: true,
  isYesterday: true,
  lastDayOfDecade: true,
  lastDayOfISOWeek: true,
  lastDayOfISOWeekYear: true,
  lastDayOfMonth: true,
  lastDayOfQuarter: true,
  lastDayOfWeek: true,
  lastDayOfYear: true,
  lightFormat: true,
  max: true,
  milliseconds: true,
  millisecondsToHours: true,
  millisecondsToMinutes: true,
  millisecondsToSeconds: true,
  min: true,
  minutesToHours: true,
  minutesToMilliseconds: true,
  minutesToSeconds: true,
  monthsToQuarters: true,
  monthsToYears: true,
  nextDay: true,
  nextFriday: true,
  nextMonday: true,
  nextSaturday: true,
  nextSunday: true,
  nextThursday: true,
  nextTuesday: true,
  nextWednesday: true,
  parse: true,
  parseISO: true,
  parseJSON: true,
  previousDay: true,
  previousFriday: true,
  previousMonday: true,
  previousSaturday: true,
  previousSunday: true,
  previousThursday: true,
  previousTuesday: true,
  previousWednesday: true,
  quartersToMonths: true,
  quartersToYears: true,
  roundToNearestMinutes: true,
  secondsToHours: true,
  secondsToMilliseconds: true,
  secondsToMinutes: true,
  set: true,
  setDate: true,
  setDay: true,
  setDayOfYear: true,
  setHours: true,
  setISODay: true,
  setISOWeek: true,
  setISOWeekYear: true,
  setMilliseconds: true,
  setMinutes: true,
  setMonth: true,
  setQuarter: true,
  setSeconds: true,
  setWeek: true,
  setWeekYear: true,
  setYear: true,
  startOfDay: true,
  startOfDecade: true,
  startOfHour: true,
  startOfISOWeek: true,
  startOfISOWeekYear: true,
  startOfMinute: true,
  startOfMonth: true,
  startOfQuarter: true,
  startOfSecond: true,
  startOfToday: true,
  startOfTomorrow: true,
  startOfWeek: true,
  startOfWeekYear: true,
  startOfYear: true,
  startOfYesterday: true,
  sub: true,
  subBusinessDays: true,
  subDays: true,
  subHours: true,
  subISOWeekYears: true,
  subMilliseconds: true,
  subMinutes: true,
  subMonths: true,
  subQuarters: true,
  subSeconds: true,
  subWeeks: true,
  subYears: true,
  toDate: true,
  weeksToDays: true,
  yearsToMonths: true,
  yearsToQuarters: true,
  af: true,
  arDZ: true,
  arMA: true,
  arSA: true,
  arTN: true,
  az: true,
  be: true,
  bg: true,
  bn: true,
  bs: true,
  ca: true,
  cs: true,
  cy: true,
  da: true,
  de: true,
  deAT: true,
  el: true,
  enAU: true,
  enCA: true,
  enGB: true,
  enIN: true,
  enNZ: true,
  enUS: true,
  enZA: true,
  eo: true,
  es: true,
  et: true,
  eu: true,
  faIR: true,
  fi: true,
  fr: true,
  frCA: true,
  gd: true,
  gl: true,
  gu: true,
  he: true,
  hi: true,
  hr: true,
  ht: true,
  hu: true,
  hy: true,
  id: true,
  is: true,
  it: true,
  ja: true,
  ka: true,
  kk: true,
  kn: true,
  ko: true,
  lb: true,
  lt: true,
  lv: true,
  mk: true,
  mn: true,
  ms: true,
  mt: true,
  nb: true,
  nl: true,
  nn: true,
  pl: true,
  pt: true,
  ptBR: true,
  ro: true,
  ru: true,
  sk: true,
  sl: true,
  sq: true,
  sr: true,
  srLatn: true,
  sv: true,
  ta: true,
  te: true,
  th: true,
  tr: true,
  ug: true,
  uk: true,
  uz: true,
  vi: true,
  zhCN: true,
  zhTW: true
};
exports.hoursToMinutes = exports.hoursToMilliseconds = exports.getYear = exports.getWeekYear = exports.getWeeksInMonth = exports.getWeekOfMonth = exports.getWeek = exports.getUnixTime = exports.getTime = exports.getSeconds = exports.getQuarter = exports.getOverlappingDaysInIntervals = exports.getMonth = exports.getMinutes = exports.getMilliseconds = exports.getISOWeekYear = exports.getISOWeeksInYear = exports.getISOWeek = exports.getISODay = exports.getHours = exports.getDecade = exports.getDaysInYear = exports.getDaysInMonth = exports.getDayOfYear = exports.getDay = exports.getDate = exports.fromUnixTime = exports.formatRFC7231 = exports.formatRFC3339 = exports.formatRelative = exports.formatISODuration = exports.formatISO9075 = exports.formatISO = exports.formatDuration = exports.formatDistanceToNowStrict = exports.formatDistanceToNow = exports.formatDistanceStrict = exports.formatDistance = exports.format = exports.endOfYesterday = exports.endOfYear = exports.endOfWeek = exports.endOfTomorrow = exports.endOfToday = exports.endOfSecond = exports.endOfQuarter = exports.endOfMonth = exports.endOfMinute = exports.endOfISOWeekYear = exports.endOfISOWeek = exports.endOfHour = exports.endOfDecade = exports.endOfDay = exports.eachYearOfInterval = exports.eachWeekOfInterval = exports.eachWeekendOfYear = exports.eachWeekendOfMonth = exports.eachWeekendOfInterval = exports.eachQuarterOfInterval = exports.eachMonthOfInterval = exports.eachMinuteOfInterval = exports.eachHourOfInterval = exports.eachDayOfInterval = exports.differenceInYears = exports.differenceInWeeks = exports.differenceInSeconds = exports.differenceInQuarters = exports.differenceInMonths = exports.differenceInMinutes = exports.differenceInMilliseconds = exports.differenceInISOWeekYears = exports.differenceInHours = exports.differenceInDays = exports.differenceInCalendarYears = exports.differenceInCalendarWeeks = exports.differenceInCalendarQuarters = exports.differenceInCalendarMonths = exports.differenceInCalendarISOWeekYears = exports.differenceInCalendarISOWeeks = exports.differenceInCalendarDays = exports.differenceInBusinessDays = exports.daysToWeeks = exports.compareDesc = exports.compareAsc = exports.closestTo = exports.closestIndexTo = exports.clamp = exports.areIntervalsOverlapping = exports.addYears = exports.addWeeks = exports.addSeconds = exports.addQuarters = exports.addMonths = exports.addMinutes = exports.addMilliseconds = exports.addISOWeekYears = exports.addHours = exports.addDays = exports.addBusinessDays = exports.add = void 0;
exports.setMonth = exports.setMinutes = exports.setMilliseconds = exports.setISOWeekYear = exports.setISOWeek = exports.setISODay = exports.setHours = exports.setDayOfYear = exports.setDay = exports.setDate = exports.set = exports.secondsToMinutes = exports.secondsToMilliseconds = exports.secondsToHours = exports.roundToNearestMinutes = exports.quartersToYears = exports.quartersToMonths = exports.previousWednesday = exports.previousTuesday = exports.previousThursday = exports.previousSunday = exports.previousSaturday = exports.previousMonday = exports.previousFriday = exports.previousDay = exports.parseJSON = exports.parseISO = exports.parse = exports.nextWednesday = exports.nextTuesday = exports.nextThursday = exports.nextSunday = exports.nextSaturday = exports.nextMonday = exports.nextFriday = exports.nextDay = exports.monthsToYears = exports.monthsToQuarters = exports.minutesToSeconds = exports.minutesToMilliseconds = exports.minutesToHours = exports.min = exports.millisecondsToSeconds = exports.millisecondsToMinutes = exports.millisecondsToHours = exports.milliseconds = exports.max = exports.lightFormat = exports.lastDayOfYear = exports.lastDayOfWeek = exports.lastDayOfQuarter = exports.lastDayOfMonth = exports.lastDayOfISOWeekYear = exports.lastDayOfISOWeek = exports.lastDayOfDecade = exports.isYesterday = exports.isWithinInterval = exports.isWeekend = exports.isWednesday = exports.isValid = exports.isTuesday = exports.isTomorrow = exports.isToday = exports.isThursday = exports.isThisYear = exports.isThisWeek = exports.isThisSecond = exports.isThisQuarter = exports.isThisMonth = exports.isThisMinute = exports.isThisISOWeek = exports.isThisHour = exports.isSunday = exports.isSaturday = exports.isSameYear = exports.isSameWeek = exports.isSameSecond = exports.isSameQuarter = exports.isSameMonth = exports.isSameMinute = exports.isSameISOWeekYear = exports.isSameISOWeek = exports.isSameHour = exports.isSameDay = exports.isPast = exports.isMonday = exports.isMatch = exports.isLeapYear = exports.isLastDayOfMonth = exports.isFuture = exports.isFriday = exports.isFirstDayOfMonth = exports.isExists = exports.isEqual = exports.isDate = exports.isBefore = exports.isAfter = exports.intlFormat = exports.intervalToDuration = exports.hoursToSeconds = void 0;
exports.ru = exports.ro = exports.ptBR = exports.pt = exports.pl = exports.nn = exports.nl = exports.nb = exports.mt = exports.ms = exports.mn = exports.mk = exports.lv = exports.lt = exports.lb = exports.ko = exports.kn = exports.kk = exports.ka = exports.ja = exports.it = exports.is = exports.id = exports.hy = exports.hu = exports.ht = exports.hr = exports.hi = exports.he = exports.gu = exports.gl = exports.gd = exports.frCA = exports.fr = exports.fi = exports.faIR = exports.eu = exports.et = exports.es = exports.eo = exports.enZA = exports.enUS = exports.enNZ = exports.enIN = exports.enGB = exports.enCA = exports.enAU = exports.el = exports.deAT = exports.de = exports.da = exports.cy = exports.cs = exports.ca = exports.bs = exports.bn = exports.bg = exports.be = exports.az = exports.arTN = exports.arSA = exports.arMA = exports.arDZ = exports.af = exports.yearsToQuarters = exports.yearsToMonths = exports.weeksToDays = exports.toDate = exports.subYears = exports.subWeeks = exports.subSeconds = exports.subQuarters = exports.subMonths = exports.subMinutes = exports.subMilliseconds = exports.subISOWeekYears = exports.subHours = exports.subDays = exports.subBusinessDays = exports.sub = exports.startOfYesterday = exports.startOfYear = exports.startOfWeekYear = exports.startOfWeek = exports.startOfTomorrow = exports.startOfToday = exports.startOfSecond = exports.startOfQuarter = exports.startOfMonth = exports.startOfMinute = exports.startOfISOWeekYear = exports.startOfISOWeek = exports.startOfHour = exports.startOfDecade = exports.startOfDay = exports.setYear = exports.setWeekYear = exports.setWeek = exports.setSeconds = exports.setQuarter = void 0;
exports.zhTW = exports.zhCN = exports.vi = exports.uz = exports.uk = exports.ug = exports.tr = exports.th = exports.te = exports.ta = exports.sv = exports.srLatn = exports.sr = exports.sq = exports.sl = exports.sk = void 0;

var _dateFns = require("date-fns");

exports.add = _dateFns.add;
exports.addBusinessDays = _dateFns.addBusinessDays;
exports.addDays = _dateFns.addDays;
exports.addHours = _dateFns.addHours;
exports.addISOWeekYears = _dateFns.addISOWeekYears;
exports.addMilliseconds = _dateFns.addMilliseconds;
exports.addMinutes = _dateFns.addMinutes;
exports.addMonths = _dateFns.addMonths;
exports.addQuarters = _dateFns.addQuarters;
exports.addSeconds = _dateFns.addSeconds;
exports.addWeeks = _dateFns.addWeeks;
exports.addYears = _dateFns.addYears;
exports.areIntervalsOverlapping = _dateFns.areIntervalsOverlapping;
exports.clamp = _dateFns.clamp;
exports.closestIndexTo = _dateFns.closestIndexTo;
exports.closestTo = _dateFns.closestTo;
exports.compareAsc = _dateFns.compareAsc;
exports.compareDesc = _dateFns.compareDesc;
exports.daysToWeeks = _dateFns.daysToWeeks;
exports.differenceInBusinessDays = _dateFns.differenceInBusinessDays;
exports.differenceInCalendarDays = _dateFns.differenceInCalendarDays;
exports.differenceInCalendarISOWeeks = _dateFns.differenceInCalendarISOWeeks;
exports.differenceInCalendarISOWeekYears = _dateFns.differenceInCalendarISOWeekYears;
exports.differenceInCalendarMonths = _dateFns.differenceInCalendarMonths;
exports.differenceInCalendarQuarters = _dateFns.differenceInCalendarQuarters;
exports.differenceInCalendarWeeks = _dateFns.differenceInCalendarWeeks;
exports.differenceInCalendarYears = _dateFns.differenceInCalendarYears;
exports.differenceInDays = _dateFns.differenceInDays;
exports.differenceInHours = _dateFns.differenceInHours;
exports.differenceInISOWeekYears = _dateFns.differenceInISOWeekYears;
exports.differenceInMilliseconds = _dateFns.differenceInMilliseconds;
exports.differenceInMinutes = _dateFns.differenceInMinutes;
exports.differenceInMonths = _dateFns.differenceInMonths;
exports.differenceInQuarters = _dateFns.differenceInQuarters;
exports.differenceInSeconds = _dateFns.differenceInSeconds;
exports.differenceInWeeks = _dateFns.differenceInWeeks;
exports.differenceInYears = _dateFns.differenceInYears;
exports.eachDayOfInterval = _dateFns.eachDayOfInterval;
exports.eachHourOfInterval = _dateFns.eachHourOfInterval;
exports.eachMinuteOfInterval = _dateFns.eachMinuteOfInterval;
exports.eachMonthOfInterval = _dateFns.eachMonthOfInterval;
exports.eachQuarterOfInterval = _dateFns.eachQuarterOfInterval;
exports.eachWeekendOfInterval = _dateFns.eachWeekendOfInterval;
exports.eachWeekendOfMonth = _dateFns.eachWeekendOfMonth;
exports.eachWeekendOfYear = _dateFns.eachWeekendOfYear;
exports.eachWeekOfInterval = _dateFns.eachWeekOfInterval;
exports.eachYearOfInterval = _dateFns.eachYearOfInterval;
exports.endOfDay = _dateFns.endOfDay;
exports.endOfDecade = _dateFns.endOfDecade;
exports.endOfHour = _dateFns.endOfHour;
exports.endOfISOWeek = _dateFns.endOfISOWeek;
exports.endOfISOWeekYear = _dateFns.endOfISOWeekYear;
exports.endOfMinute = _dateFns.endOfMinute;
exports.endOfMonth = _dateFns.endOfMonth;
exports.endOfQuarter = _dateFns.endOfQuarter;
exports.endOfSecond = _dateFns.endOfSecond;
exports.endOfToday = _dateFns.endOfToday;
exports.endOfTomorrow = _dateFns.endOfTomorrow;
exports.endOfWeek = _dateFns.endOfWeek;
exports.endOfYear = _dateFns.endOfYear;
exports.endOfYesterday = _dateFns.endOfYesterday;
exports.format = _dateFns.format;
exports.formatDistance = _dateFns.formatDistance;
exports.formatDistanceStrict = _dateFns.formatDistanceStrict;
exports.formatDistanceToNow = _dateFns.formatDistanceToNow;
exports.formatDistanceToNowStrict = _dateFns.formatDistanceToNowStrict;
exports.formatDuration = _dateFns.formatDuration;
exports.formatISO = _dateFns.formatISO;
exports.formatISO9075 = _dateFns.formatISO9075;
exports.formatISODuration = _dateFns.formatISODuration;
exports.formatRelative = _dateFns.formatRelative;
exports.formatRFC3339 = _dateFns.formatRFC3339;
exports.formatRFC7231 = _dateFns.formatRFC7231;
exports.fromUnixTime = _dateFns.fromUnixTime;
exports.getDate = _dateFns.getDate;
exports.getDay = _dateFns.getDay;
exports.getDayOfYear = _dateFns.getDayOfYear;
exports.getDaysInMonth = _dateFns.getDaysInMonth;
exports.getDaysInYear = _dateFns.getDaysInYear;
exports.getDecade = _dateFns.getDecade;
exports.getHours = _dateFns.getHours;
exports.getISODay = _dateFns.getISODay;
exports.getISOWeek = _dateFns.getISOWeek;
exports.getISOWeeksInYear = _dateFns.getISOWeeksInYear;
exports.getISOWeekYear = _dateFns.getISOWeekYear;
exports.getMilliseconds = _dateFns.getMilliseconds;
exports.getMinutes = _dateFns.getMinutes;
exports.getMonth = _dateFns.getMonth;
exports.getOverlappingDaysInIntervals = _dateFns.getOverlappingDaysInIntervals;
exports.getQuarter = _dateFns.getQuarter;
exports.getSeconds = _dateFns.getSeconds;
exports.getTime = _dateFns.getTime;
exports.getUnixTime = _dateFns.getUnixTime;
exports.getWeek = _dateFns.getWeek;
exports.getWeekOfMonth = _dateFns.getWeekOfMonth;
exports.getWeeksInMonth = _dateFns.getWeeksInMonth;
exports.getWeekYear = _dateFns.getWeekYear;
exports.getYear = _dateFns.getYear;
exports.hoursToMilliseconds = _dateFns.hoursToMilliseconds;
exports.hoursToMinutes = _dateFns.hoursToMinutes;
exports.hoursToSeconds = _dateFns.hoursToSeconds;
exports.intervalToDuration = _dateFns.intervalToDuration;
exports.intlFormat = _dateFns.intlFormat;
exports.isAfter = _dateFns.isAfter;
exports.isBefore = _dateFns.isBefore;
exports.isDate = _dateFns.isDate;
exports.isEqual = _dateFns.isEqual;
exports.isExists = _dateFns.isExists;
exports.isFirstDayOfMonth = _dateFns.isFirstDayOfMonth;
exports.isFriday = _dateFns.isFriday;
exports.isFuture = _dateFns.isFuture;
exports.isLastDayOfMonth = _dateFns.isLastDayOfMonth;
exports.isLeapYear = _dateFns.isLeapYear;
exports.isMatch = _dateFns.isMatch;
exports.isMonday = _dateFns.isMonday;
exports.isPast = _dateFns.isPast;
exports.isSameDay = _dateFns.isSameDay;
exports.isSameHour = _dateFns.isSameHour;
exports.isSameISOWeek = _dateFns.isSameISOWeek;
exports.isSameISOWeekYear = _dateFns.isSameISOWeekYear;
exports.isSameMinute = _dateFns.isSameMinute;
exports.isSameMonth = _dateFns.isSameMonth;
exports.isSameQuarter = _dateFns.isSameQuarter;
exports.isSameSecond = _dateFns.isSameSecond;
exports.isSameWeek = _dateFns.isSameWeek;
exports.isSameYear = _dateFns.isSameYear;
exports.isSaturday = _dateFns.isSaturday;
exports.isSunday = _dateFns.isSunday;
exports.isThisHour = _dateFns.isThisHour;
exports.isThisISOWeek = _dateFns.isThisISOWeek;
exports.isThisMinute = _dateFns.isThisMinute;
exports.isThisMonth = _dateFns.isThisMonth;
exports.isThisQuarter = _dateFns.isThisQuarter;
exports.isThisSecond = _dateFns.isThisSecond;
exports.isThisWeek = _dateFns.isThisWeek;
exports.isThisYear = _dateFns.isThisYear;
exports.isThursday = _dateFns.isThursday;
exports.isToday = _dateFns.isToday;
exports.isTomorrow = _dateFns.isTomorrow;
exports.isTuesday = _dateFns.isTuesday;
exports.isValid = _dateFns.isValid;
exports.isWednesday = _dateFns.isWednesday;
exports.isWeekend = _dateFns.isWeekend;
exports.isWithinInterval = _dateFns.isWithinInterval;
exports.isYesterday = _dateFns.isYesterday;
exports.lastDayOfDecade = _dateFns.lastDayOfDecade;
exports.lastDayOfISOWeek = _dateFns.lastDayOfISOWeek;
exports.lastDayOfISOWeekYear = _dateFns.lastDayOfISOWeekYear;
exports.lastDayOfMonth = _dateFns.lastDayOfMonth;
exports.lastDayOfQuarter = _dateFns.lastDayOfQuarter;
exports.lastDayOfWeek = _dateFns.lastDayOfWeek;
exports.lastDayOfYear = _dateFns.lastDayOfYear;
exports.lightFormat = _dateFns.lightFormat;
exports.max = _dateFns.max;
exports.milliseconds = _dateFns.milliseconds;
exports.millisecondsToHours = _dateFns.millisecondsToHours;
exports.millisecondsToMinutes = _dateFns.millisecondsToMinutes;
exports.millisecondsToSeconds = _dateFns.millisecondsToSeconds;
exports.min = _dateFns.min;
exports.minutesToHours = _dateFns.minutesToHours;
exports.minutesToMilliseconds = _dateFns.minutesToMilliseconds;
exports.minutesToSeconds = _dateFns.minutesToSeconds;
exports.monthsToQuarters = _dateFns.monthsToQuarters;
exports.monthsToYears = _dateFns.monthsToYears;
exports.nextDay = _dateFns.nextDay;
exports.nextFriday = _dateFns.nextFriday;
exports.nextMonday = _dateFns.nextMonday;
exports.nextSaturday = _dateFns.nextSaturday;
exports.nextSunday = _dateFns.nextSunday;
exports.nextThursday = _dateFns.nextThursday;
exports.nextTuesday = _dateFns.nextTuesday;
exports.nextWednesday = _dateFns.nextWednesday;
exports.parse = _dateFns.parse;
exports.parseISO = _dateFns.parseISO;
exports.parseJSON = _dateFns.parseJSON;
exports.previousDay = _dateFns.previousDay;
exports.previousFriday = _dateFns.previousFriday;
exports.previousMonday = _dateFns.previousMonday;
exports.previousSaturday = _dateFns.previousSaturday;
exports.previousSunday = _dateFns.previousSunday;
exports.previousThursday = _dateFns.previousThursday;
exports.previousTuesday = _dateFns.previousTuesday;
exports.previousWednesday = _dateFns.previousWednesday;
exports.quartersToMonths = _dateFns.quartersToMonths;
exports.quartersToYears = _dateFns.quartersToYears;
exports.roundToNearestMinutes = _dateFns.roundToNearestMinutes;
exports.secondsToHours = _dateFns.secondsToHours;
exports.secondsToMilliseconds = _dateFns.secondsToMilliseconds;
exports.secondsToMinutes = _dateFns.secondsToMinutes;
exports.set = _dateFns.set;
exports.setDate = _dateFns.setDate;
exports.setDay = _dateFns.setDay;
exports.setDayOfYear = _dateFns.setDayOfYear;
exports.setHours = _dateFns.setHours;
exports.setISODay = _dateFns.setISODay;
exports.setISOWeek = _dateFns.setISOWeek;
exports.setISOWeekYear = _dateFns.setISOWeekYear;
exports.setMilliseconds = _dateFns.setMilliseconds;
exports.setMinutes = _dateFns.setMinutes;
exports.setMonth = _dateFns.setMonth;
exports.setQuarter = _dateFns.setQuarter;
exports.setSeconds = _dateFns.setSeconds;
exports.setWeek = _dateFns.setWeek;
exports.setWeekYear = _dateFns.setWeekYear;
exports.setYear = _dateFns.setYear;
exports.startOfDay = _dateFns.startOfDay;
exports.startOfDecade = _dateFns.startOfDecade;
exports.startOfHour = _dateFns.startOfHour;
exports.startOfISOWeek = _dateFns.startOfISOWeek;
exports.startOfISOWeekYear = _dateFns.startOfISOWeekYear;
exports.startOfMinute = _dateFns.startOfMinute;
exports.startOfMonth = _dateFns.startOfMonth;
exports.startOfQuarter = _dateFns.startOfQuarter;
exports.startOfSecond = _dateFns.startOfSecond;
exports.startOfToday = _dateFns.startOfToday;
exports.startOfTomorrow = _dateFns.startOfTomorrow;
exports.startOfWeek = _dateFns.startOfWeek;
exports.startOfWeekYear = _dateFns.startOfWeekYear;
exports.startOfYear = _dateFns.startOfYear;
exports.startOfYesterday = _dateFns.startOfYesterday;
exports.sub = _dateFns.sub;
exports.subBusinessDays = _dateFns.subBusinessDays;
exports.subDays = _dateFns.subDays;
exports.subHours = _dateFns.subHours;
exports.subISOWeekYears = _dateFns.subISOWeekYears;
exports.subMilliseconds = _dateFns.subMilliseconds;
exports.subMinutes = _dateFns.subMinutes;
exports.subMonths = _dateFns.subMonths;
exports.subQuarters = _dateFns.subQuarters;
exports.subSeconds = _dateFns.subSeconds;
exports.subWeeks = _dateFns.subWeeks;
exports.subYears = _dateFns.subYears;
exports.toDate = _dateFns.toDate;
exports.weeksToDays = _dateFns.weeksToDays;
exports.yearsToMonths = _dateFns.yearsToMonths;
exports.yearsToQuarters = _dateFns.yearsToQuarters;

var _locale = require("date-fns/locale");

exports.af = _locale.af;
exports.arDZ = _locale.arDZ;
exports.arMA = _locale.arMA;
exports.arSA = _locale.arSA;
exports.arTN = _locale.arTN;
exports.az = _locale.az;
exports.be = _locale.be;
exports.bg = _locale.bg;
exports.bn = _locale.bn;
exports.bs = _locale.bs;
exports.ca = _locale.ca;
exports.cs = _locale.cs;
exports.cy = _locale.cy;
exports.da = _locale.da;
exports.de = _locale.de;
exports.deAT = _locale.deAT;
exports.el = _locale.el;
exports.enAU = _locale.enAU;
exports.enCA = _locale.enCA;
exports.enGB = _locale.enGB;
exports.enIN = _locale.enIN;
exports.enNZ = _locale.enNZ;
exports.enUS = _locale.enUS;
exports.enZA = _locale.enZA;
exports.eo = _locale.eo;
exports.es = _locale.es;
exports.et = _locale.et;
exports.eu = _locale.eu;
exports.faIR = _locale.faIR;
exports.fi = _locale.fi;
exports.fr = _locale.fr;
exports.frCA = _locale.frCA;
exports.gd = _locale.gd;
exports.gl = _locale.gl;
exports.gu = _locale.gu;
exports.he = _locale.he;
exports.hi = _locale.hi;
exports.hr = _locale.hr;
exports.ht = _locale.ht;
exports.hu = _locale.hu;
exports.hy = _locale.hy;
exports.id = _locale.id;
exports.is = _locale.is;
exports.it = _locale.it;
exports.ja = _locale.ja;
exports.ka = _locale.ka;
exports.kk = _locale.kk;
exports.kn = _locale.kn;
exports.ko = _locale.ko;
exports.lb = _locale.lb;
exports.lt = _locale.lt;
exports.lv = _locale.lv;
exports.mk = _locale.mk;
exports.mn = _locale.mn;
exports.ms = _locale.ms;
exports.mt = _locale.mt;
exports.nb = _locale.nb;
exports.nl = _locale.nl;
exports.nn = _locale.nn;
exports.pl = _locale.pl;
exports.pt = _locale.pt;
exports.ptBR = _locale.ptBR;
exports.ro = _locale.ro;
exports.ru = _locale.ru;
exports.sk = _locale.sk;
exports.sl = _locale.sl;
exports.sq = _locale.sq;
exports.sr = _locale.sr;
exports.srLatn = _locale.srLatn;
exports.sv = _locale.sv;
exports.ta = _locale.ta;
exports.te = _locale.te;
exports.th = _locale.th;
exports.tr = _locale.tr;
exports.ug = _locale.ug;
exports.uk = _locale.uk;
exports.uz = _locale.uz;
exports.vi = _locale.vi;
exports.zhCN = _locale.zhCN;
exports.zhTW = _locale.zhTW;

var _anyToDate = require("./anyToDate");

Object.keys(_anyToDate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _anyToDate[key]) return;
  exports[key] = _anyToDate[key];
});

var _formatDate = require("./formatDate");

Object.keys(_formatDate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formatDate[key]) return;
  exports[key] = _formatDate[key];
});

var _formatDistanceAgo = require("./formatDistanceAgo");

Object.keys(_formatDistanceAgo).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _formatDistanceAgo[key]) return;
  exports[key] = _formatDistanceAgo[key];
});

var _intervalToRestrictiveDuration = require("./intervalToRestrictiveDuration");

Object.keys(_intervalToRestrictiveDuration).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _intervalToRestrictiveDuration[key]) return;
  exports[key] = _intervalToRestrictiveDuration[key];
});

var _numeralDayToChineseDay = require("./numeralDayToChineseDay");

Object.keys(_numeralDayToChineseDay).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _numeralDayToChineseDay[key]) return;
  exports[key] = _numeralDayToChineseDay[key];
});
