// vtils/_cjs/react/createGlobalState.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.createGlobalState = createGlobalState;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = require("react");

var _utils = require("../utils");

function createGlobalState(_initialState, _customResult) {
  var initialState;
  var customResult;

  if (typeof _customResult === 'function') {
    initialState = _initialState;
    customResult = _customResult;
  } else if (typeof _initialState === 'function') {
    initialState = undefined;
    customResult = _initialState;
  } else {
    initialState = _initialState;
    customResult = undefined;
  }

  var bus = new _utils.EventBus();
  var currentGlobalState = initialState;

  var getGlobalState = function getGlobalState() {
    return currentGlobalState;
  };

  var setGlobalState = function setGlobalState(nextGlobalState) {
    if (typeof nextGlobalState === 'function') {
      nextGlobalState = nextGlobalState(currentGlobalState);
    }

    if (nextGlobalState !== currentGlobalState) {
      var _prevGlobalState = currentGlobalState;
      currentGlobalState = nextGlobalState;
      bus.emit('setGlobalState', nextGlobalState, _prevGlobalState);
    }
  };

  var setGlobalStatePartial = function setGlobalStatePartial(nextGlobalState) {
    if (typeof nextGlobalState === 'function') {
      nextGlobalState = nextGlobalState(currentGlobalState);
    }

    nextGlobalState = (0, _extends2.default)({}, currentGlobalState, nextGlobalState);
    var prevGlobalState = currentGlobalState;
    currentGlobalState = nextGlobalState;
    bus.emit('setGlobalState', nextGlobalState, prevGlobalState);
  };

  var watchGlobalState = function watchGlobalState(callback) {
    return bus.on('setGlobalState', callback);
  };

  var watchGlobalStateImmediate = function watchGlobalStateImmediate(callback) {
    callback(initialState, undefined);
    return bus.on('setGlobalState', callback);
  };

  var useCustomResult = typeof customResult === 'function';

  var useGlobalState = function useGlobalState() {
    var _useState = (0, _react.useState)(currentGlobalState),
        state = _useState[0],
        setState = _useState[1];

    (0, _react.useEffect)(function () {
      return watchGlobalState(setState);
    }, []);
    return useCustomResult ? customResult({
      state: state,
      setState: setGlobalState
    }) : [state, setGlobalState];
  };

  useGlobalState.getState = getGlobalState;
  useGlobalState.setState = setGlobalState;
  useGlobalState.setStatePartial = setGlobalStatePartial;
  useGlobalState.watchState = watchGlobalState;
  useGlobalState.watchStateImmediate = watchGlobalStateImmediate;
  return useGlobalState;
}
