// vtils/_cjs/react/useValidator.js
"use strict";

exports.__esModule = true;
exports.useValidator = useValidator;

var _validator = require("../validator");

var _react = require("react");

var validateOptions = {
  strict: true,
  abortEarly: true,
  stripUnknown: false,
  recursive: true
};

function useValidator(data, dataDeps, schema, schemaDeps) {
  var _dataDeps;

  var _schema;

  var _schemaDeps;

  if (Array.isArray(dataDeps)) {
    _dataDeps = dataDeps;
    _schema = schema;
    _schemaDeps = schemaDeps || [];
  } else {
    _dataDeps = [];
    _schema = dataDeps;
    _schemaDeps = schema || [];
  }

  var yupSchema = (0, _react.useMemo)(function () {
    var schemaOrObjectDefinitions = _schema(_validator.yup);

    if (schemaOrObjectDefinitions.__isYupSchema__) {
      return schemaOrObjectDefinitions;
    }

    return _validator.yup.object(schemaOrObjectDefinitions);
  }, _schemaDeps);
  var result = (0, _react.useMemo)(function () {
    var r = yupSchema.validatePlusSync(data, validateOptions);
    return {
      data: r.data,
      valid: !r.error,
      error: r.error
    };
  }, [].concat(_dataDeps, [yupSchema]));
  return result;
}
