// vtils/_cjs/mp/miniProgramBus.js
"use strict";

exports.__esModule = true;
exports.miniProgramBus = exports.currentPageListeners = exports.pageListeners = exports.pageListenerToCurrentPageListener = void 0;

var _utils = require("../utils");

var _patchMiniProgram = require("./patchMiniProgram");

/** @private */
var pageListenerToCurrentPageListener = {
  pageShow: 'currentPageShow',
  pageHide: 'currentPageHide',
  pageReady: 'currentPageReady',
  pageUnload: 'currentPageUnload',
  pagePullDownRefresh: 'currentPagePullDownRefresh',
  pageReachBottom: 'currentPageReachBottom',
  pageShareAppMessage: 'currentPageShareAppMessage',
  pageShareTimeline: 'currentPageShareTimeline',
  pageAddToFavorites: 'currentPageAddToFavorites',
  pageResize: 'currentPageResize',
  pageTabItemTap: 'currentPageTabItemTap'
};
/** @private */

exports.pageListenerToCurrentPageListener = pageListenerToCurrentPageListener;
var pageListeners = Object.keys(pageListenerToCurrentPageListener);
/** @private */

exports.pageListeners = pageListeners;
var currentPageListeners = pageListeners.map(function (pageListener) {
  return pageListenerToCurrentPageListener[pageListener];
}); // @ts-ignore

exports.currentPageListeners = currentPageListeners;
var miniProgramBus = new _utils.EventBus({
  beforeOn: currentPageListeners.reduce(function (res, currentPageListenerName) {
    res[currentPageListenerName] = function (cb) {
      cb.__EVENT_BUS_TAG__ = _patchMiniProgram.patchMiniProgram.__CURRENT_PAGE_ID__;
      return cb;
    };

    return res;
  }, {})
});
exports.miniProgramBus = miniProgramBus;
