// vtils/_cjs/mp/submit.js
"use strict";

exports.__esModule = true;
exports.submit = void 0;

var _utils = require("../utils");

var _ensureInMiniProgram = require("./ensureInMiniProgram");

/**
 * 对提交类行为的封装。
 */
var submit = (0, _utils.createSubmit)({
  start: function start(message) {
    (0, _ensureInMiniProgram.ensureInMiniProgram)(function (mp) {
      mp.showLoading({
        title: message || '',
        mask: true
      });
    });
  },
  fail: function fail(message, duration) {
    (0, _ensureInMiniProgram.ensureInMiniProgram)(function (mp) {
      mp.hideLoading();
      mp.showToast({
        title: message,
        icon: 'none',
        duration: duration
      });
    });
  },
  success: function success(message, duration) {
    (0, _ensureInMiniProgram.ensureInMiniProgram)(function (mp) {
      mp.hideLoading();
      mp.showToast({
        title: message,
        icon: 'success',
        duration: duration
      });
    });
  },
  complete: function complete() {
    (0, _ensureInMiniProgram.ensureInMiniProgram)(function (mp) {
      mp.hideLoading();
    });
  }
});
exports.submit = submit;
